### PR TITLE
Defanged dummy type arguments.

### PR DESCRIPTION
**Pull request by paulp**
Spending a few hours on yet another tpe vs. tpeHK bug drove me
to finally go after the dragon in his cave. I created a separate
method for when you want to receive an invalid type which will
cause mysterious crashes later if it is out of your sight for five
seconds.

  def tpeHK : Type // unapplied type params stay unapplied
  def tpe_\* : Type // unapplied type params applied as dummy args
  def tpe   : Type // if there are unapplied type params, ABORT

Actually it doesn't really abort, but I had it print a stack trace
during development so I could track down every site which blindly
calls tpe and see why they were doing it. Now it only admonishes
you under -Ydebug.

This way is a big step forward: you have to make a choice, which
is good, because if you choose not to decide you still have made a
choice, only mostly what you have chosen is bugs.

### CODE REVIEW

**Comment by paulp** [comment]
Review by @retronym.

File: src/continuations/plugin/scala/tools/selectivecps/SelectiveCPSTransform.scala [hunk 1]
```diff
@@ -56,7 +56,7 @@ abstract class SelectiveCPSTransform extends PluginComponent with
       case _ =>
         getExternalAnswerTypeAnn(tp) match {
           case Some((res, outer)) =>
-            appliedType(Context.tpe, List(removeAllCPSAnnotations(tp), res, outer))
+            appliedType(Context.tpeHK, List(removeAllCPSAnnotations(tp), res, outer))
```
Comment (retronym) [inline src/continuations/plugin/scala/tools/selectivecps/SelectiveCPSTransform.scala:59]: It seems pretty loose that `appliedType` admits and discards the dummies:

``` scala
case TypeRef(pre, sym, _)                           => copyTypeRef(tycon, pre, sym, args)
```

You might pop another warning in there to see what else tries to get through.

File: src/reflect/scala/reflect/internal/Symbols.scala [hunk 2]
```diff
@@ -2625,20 +2642,30 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
      *    tsym.tpe  = TypeRef(NoPrefix, T, List())
      *  }}}
      */
-    override def tpe: Type = {
+    override def tpe: Type    = tpeOfKind(kindStar = false)
+    override def tpe_* : Type = tpeOfKind(kindStar = true)
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Symbols.scala]: Sometimes all you need is a good name. I like this one.

File: src/reflect/scala/reflect/internal/transform/Erasure.scala [hunk 3]
```diff
@@ -69,7 +69,7 @@ trait Erasure {
   //
   // This requires that cls.isClass.
   protected def rebindInnerClass(pre: Type, cls: Symbol): Type = {
-    if (cls.owner.isClass) cls.owner.tpe else pre // why not cls.isNestedClass?
+    if (cls.owner.isClass) cls.owner.thisType else pre // why not cls.isNestedClass?
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/transform/Erasure.scala]: For my benefit, can you explain this one slowly?

File: src/reflect/scala/reflect/internal/Symbols.scala [hunk 4]
```diff
@@ -1165,17 +1165,34 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
...
-     *    @M you should use tpeHK for a type symbol with type parameters if
-     *       the kind of the type need not be *, as tpe introduces dummy arguments
-     *       to generate a type of kind *
-     *  for a term symbol, its usual type.
+    /** Get type. The type of a term symbol is its usual type.
+     *  For a type symbol with no type parameters, it is the type corresponding to the symbol
+     *  itself. For a type symbol with type parameters, you
+     *  must determine whether you want tpeHK or tpe_*.  If you
+     *  call .tpe on such a symbol, you may trigger an assertion.
      *  See the tpe/tpeHK overrides in TypeSymbol for more.
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Symbols.scala]: I think the documentation in the `TypeSymbol` should be pulled up to this level and merged with these comments.

**Comment by retronym** [comment]
Any observable improvements in behaviour? Or purely avoiding the new warning?

Just to be clear: your immediate proposal is to leave most of the existing `.tpe` calls, and to call on `-Ydebug` to pinpoint the buggy usages when diagnosing crashers. 

You could also escalate it to an error with another flag and use that in a paranoid run of partest and the community build.

**Comment by paulp** [comment]
"your immediate proposal is to leave most of the existing .tpe calls, and to call on -Ydebug to pinpoint the buggy usages when diagnosing crashers."

None of the remaining .tpe calls ever call .tpe in a situation where dummies would come out, when compiling scala or running the test suite.  Of course a few more will when we branch out or go community (let's see what happens with scalaz) but mostly they will not be participants in these kinds of bugs.  Going forward I'm amenable to any number of arrangements.  Right now changing more than necessary, especially on something which hasn't been approved, can only buy me merge pain. 

My proposal at this second is only to get this foot in the door.

**Comment by retronym** [comment]
It's an excellent first step. LGTM. But I believe that @adriaanm is due back, so we should double check that it LGTH, too.

**Comment by paulp** [comment]
More reviewing by @adriaanm

**Comment by gkossakowski** [comment]
PLS REBUILD ALL.

@adriaanm will be back at work on Monday so I'll let it hang until then.

File: src/reflect/scala/reflect/internal/Symbols.scala [hunk 5]
```diff
@@ -2625,20 +2642,30 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
...
           if (isInitialized) tpePeriod = currentPeriod
           tpeCache = NoType
           val targs =
-            if (phase.erasedTypes && this != ArrayClass) List()
-            else unsafeTypeParams map (_.typeConstructor)
-            //@M! use typeConstructor to generate dummy type arguments,
-            // sym.tpe should not be called on a symbol that's supposed to be a higher-kinded type
-            // memberType should be used instead, that's why it uses tpeHK and not tpe
+            if (phase.erasedTypes && this != ArrayClass) Nil
+            else {
```
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Symbols.scala]: couldn't this else's block be replaced by the following?

```
{
  val dummies = unsafeTypeParams map (_.typeConstructor)

  if (settings.debug.value && dummies.nonEmpty && !kindStar)
    printCaller(s"""Call to ${this.tpe} with unapplied ${dummies mkString ", "}: call tpe_* or tpeHK""")(dummies)

  dummies
}
```

File: src/reflect/scala/reflect/internal/Symbols.scala [hunk 6]
```diff
@@ -1165,17 +1165,34 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
...
+     *  call .tpe on such a symbol, you may trigger an assertion.
      *  See the tpe/tpeHK overrides in TypeSymbol for more.
      */
     def tpe: Type = info
+
+    /** The type of this symbol, with any unapplied type parameters
+     *  remaining unapplied (which means the type constructor.)
+     *  It may be of any kind.
+     */
     def tpeHK: Type = tpe
```
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/Symbols.scala]: i'd vote for renaming `tpeHK` to `typeConstructor` while we're at it

File: src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala [hunk 7]
```diff
@@ -83,7 +83,7 @@ abstract class GenASM extends SubComponent with BytecodeWriters {
         // Before erasure so we can identify generic mains.
         enteringErasure {
           val companion     = sym.linkedClassOfClass
-          val companionMain = companion.tpe.member(nme.main)
+          val companionMain = companion.thisType.member(nme.main)
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala]: i'm not sure we should change this as part of this refactoring -- it seems unrelated

File: src/compiler/scala/tools/nsc/transform/ExplicitOuter.scala [hunk 8]
```diff
@@ -550,13 +544,6 @@ abstract class ExplicitOuter extends InfoTransform
           }
 
         case _ =>
-          if (settings.Xmigration28.value) tree match {
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/ExplicitOuter.scala:553]: i know it's a pain to submit several pull requests, but let's not couple the fate of sure-win cleanups and more ambitious refactorings

File: src/continuations/plugin/scala/tools/selectivecps/CPSUtils.scala [hunk 9]
```diff
@@ -57,7 +57,7 @@ trait CPSUtils {
   protected def newMarker(sym: Symbol): AnnotationInfo = AnnotationInfo marker sym.tpe
 
   protected def newCpsParamsMarker(tp1: Type, tp2: Type) =
-    newMarker(appliedType(MarkerCPSTypes.tpe, List(tp1, tp2)))
+    newMarker(appliedType(MarkerCPSTypes, tp1, tp2))
```
Comment (adriaanm) [inline src/continuations/plugin/scala/tools/selectivecps/CPSUtils.scala:60]: I like this approach. It reminds me of how I wanted to scrap `tpe`/`tpeHK`(and `tpe_*`) on `Symbol` altogether (leave `info`, of course) and provide a smarter `TypeRef` constructor (perhaps one that takes an expected kind). This overload of `appliedType` that takes a symbol and a variable-length list of type arguments would be the TypeRef constructor (supplying dummies when necessary, emitting a warning for non-empty dummies when in debug mode).

The other would be `TypeConstructorRef`.

Both should have overloads that take a Type (the prefix) as their first argument.

**Comment by paulp** [comment]
Since everything is supposed to go through the typeRef creator, we could do that by way of def typeConstructorRef.  All solutions are roughly equivalent in my eyes as long as they give us a means to express what is intended at the point of creation and a means to recognize it when what is happening does not match what is intended.

File: src/reflect/scala/reflect/internal/Symbols.scala (hunk 6 above)
Comment (paulp) [inline src/reflect/scala/reflect/internal/Symbols.scala]: Can you think of a name which is shorter and/or has higher symmetry with tpe_\* ? Or, I'm not that attached to tpe_\* because of the clumsiness of _ containing symbol names, and I think it loses a lot of appeal if it isn't coupled with tpeHK. I'd like the names to be visibly complementary.

Comment (paulp) [inline src/reflect/scala/reflect/internal/Symbols.scala]: Ha ha, github interpreted my

```
tpe_*
```

uses as emphasis between the underscores.

File: src/reflect/scala/reflect/internal/Symbols.scala (hunk 5 above)
Comment (paulp) [inline src/reflect/scala/reflect/internal/Symbols.scala]: Yeah, suboptimal simplification from more complicated interim state.

**Comment by paulp** [comment]
I broke it up into little tiny separate commits, and incorporated most of what I heard.  I didn't rename tpeHK because I figure we can do that anytime and should make sure we all agree what we're doing.  (I also noticed that typeConstructor aborts if you call it on a TermSymbol, but tpeHK doesn't and maybe shouldn't; so  I left it as the non-exception-throwing alternative.)

Re-review by @adriaanm and @retronym .  There's a little bit of new material, nothing too thrilling.

**Comment by paulp** [comment]
BUMP! @adriaanm @retronym a rubber stamping is fine, I know it's good.

**Comment by retronym** [comment]
Oh, and this LGTM.

**Comment by paulp** [comment]
THANK THE STARS
